### PR TITLE
Allow service ports to be specified as a (host, port) instead

### PR DESCRIPTION
**Pull request by bmerry**
My main use case is to pass `diagnostics_port=(host, port)` to
LocalCluster to run a world-accessible Bokeh server even when the
scheduler uses inproc://. However, the design should be more general
e.g. it is a step towards dask/distributed#1687.

I haven't written any tests or documented it yet - I'd like to get
feedback on whether this is the right approach before investing in
those.

### CODE REVIEW

**Comment by mrocklin** [comment]
In principle this seems reasonable to me.

**Comment by mrocklin** [comment]
My apologies for the delay in response.  I'm glad that you're taking this on.

**Comment by mrocklin** [comment]
It would be good to see things like tests and doc changes around this code change.

**Comment by bmerry** [comment]
> It would be good to see things like tests and doc changes around this code change.

Sure, I'll get to that next now that you're said you're happy in principle.

I haven't started looking at the testing framework yet. Can you suggest which area I should look at to add tests?

**Comment by bmerry** [comment]
I've added documentation for the `LocalCluster` use case. I'm very new to distributed and haven't read most of the docs, so I'm not sure if there are other places the documentation needs to be extended.

I added a test as well. It's a bit of a copy-paste job from the test immediately above it. Let me know if you'd prefer me to try to combine them into one parametrized test.

File: distributed/tests/test_scheduler.py [hunk 1]
```diff
@@ -1068,6 +1068,30 @@ def test_service_hosts_match_scheduler():
...
+def test_service_custom_host():
+    pytest.importorskip('bokeh')
+    from distributed.bokeh.scheduler import BokehScheduler
+    services = {('bokeh', ('127.0.0.3', 0)): BokehScheduler}
+
+    s = Scheduler(services=services)
+    yield s.start('tcp://0.0.0.0')
+
+    sock = first(s.services['bokeh'].server._http._sockets.values())
+    assert sock.getsockname()[0] in ('::', '0.0.0.0')
```
Comment (mrocklin) [inline distributed/tests/test_scheduler.py]: I'm curious, why wouldn't this be the hostname that we specified above?  (My understanding of networking here is poor)

File: distributed/tests/test_scheduler.py [hunk 2]
```diff
@@ -1068,6 +1068,30 @@ def test_service_hosts_match_scheduler():
         yield s.close()
 
 
+@pytest.mark.skipif(not sys.platform.startswith('linux'),
+                    reason="Need 127.0.0.* to mean localhost")
+@gen_test(timeout=None)
```
Comment (mrocklin) [inline distributed/tests/test_scheduler.py]: Can we remove `timeout=None` to re-enable the standard 10s timeout?

File: distributed/tests/test_scheduler.py (hunk 1 above)
Comment (bmerry) [reply distributed/tests/test_scheduler.py]: You're right, it should, and it's confirmed by the Travis failure.

It seems that when I run `py.test distributed --verbose` it's silently exiting during `distributed/deploy/tests/test_ssh.py::test_cluster` without indicating any sort of failure, and hadn't even reached my new test. So I was falsely assuming it was passing.

File: distributed/tests/test_scheduler.py (hunk 2 above)
Comment (bmerry) [reply distributed/tests/test_scheduler.py]: I can - it was copy-pasted from the test above it, which is very similar. Should I remove it from both tests?

Comment (mrocklin) [reply distributed/tests/test_scheduler.py]: Yes, that sounds good

File: distributed/tests/test_scheduler.py [hunk 3]
```diff
@@ -1068,6 +1068,30 @@ def test_service_hosts_match_scheduler():
...
+    assert sock.getsockname()[0] == '127.0.0.3'
+    yield s.close()
+
+    for host in ['tcp://127.0.0.2', 'tcp://127.0.0.2:38275']:
+        s = Scheduler(services=services)
+        yield s.start(host)
+
+        sock = first(s.services['bokeh'].server._http._sockets.values())
+        assert sock.getsockname()[0] == '127.0.0.3'
+        yield s.close()
```
Comment (mrocklin) [inline distributed/tests/test_scheduler.py]: You might also be able to clean up this test a bit, perhaps by merging the for loop with the example above?  Also, does it make sense to try with each of these scheduler start hosts in particular?  I get the copy-paste history, but we should also do some thinking about cases that might be more relevant to your situation.

Comment (bmerry) [reply distributed/tests/test_scheduler.py]: I've given it another go, merging it with the test from which it was copy-pasting, and also reducing the amount of copy-paste in that test while I'm at it.

**Comment by mrocklin** [comment]
This seems good to me.  Merging.  Thanks for the analysis and for the fix @bmerry !

**Pull request merged**